pkg/zstdchunked/compression.go

```diff
@@ -0,0 +1,477 @@
+package zstdchunked
+
+import (
+	"bytes"
+	"encoding/base64"
+	"encoding/binary"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"time"
+
+	"github.com/containers/storage/pkg/ioutils"
+	"github.com/klauspost/compress/zstd"
+	digest "github.com/opencontainers/go-digest"
+	"github.com/vbatts/tar-split/archive/tar"
+)
+
+type TOC struct {
+	Version int            `json:"version"`
+	Entries []FileMetadata `json:"entries"`
+}
+
+type FileMetadata struct {
+	Type       string            `json:"type"`
+	Name       string            `json:"name"`
+	Linkname   string            `json:"linkName,omitempty"`
+	Mode       int64             `json:"mode,omitempty"`
+	Size       int64             `json:"size"`
+	UID        int               `json:"uid"`
+	GID        int               `json:"gid"`
+	ModTime    time.Time         `json:"modtime"`
+	AccessTime time.Time         `json:"accesstime"`
+	ChangeTime time.Time         `json:"changetime"`
+	Devmajor   int64             `json:"devMajor"`
+	Devminor   int64             `json:"devMinor"`
+	Xattrs     map[string]string `json:"xattrs,omitempty"`
+	Digest     string            `json:"digest,omitempty"`
+	Offset     int64             `json:"offset,omitempty"`
+	EndOffset  int64             `json:"endOffset,omitempty"`
+}
```

Comment:
Why compression per file (not smaller chunk)? Accessing a small part of a big file ends up downloading the entire file contents?

Reply:
yes this is something not supported yet

How does estargz do it?  Rolling checksum on the single file or does it split in a different way?

Reply:
We have [`chunk` type of TOC Entry](https://github.com/containerd/stargz-snapshotter/blob/v0.4.1/docs/stargz-estargz.md#the-structure). This stores the chunk's offset(absolute offset and the one relative to the top of the file), size and digest. https://github.com/containerd/stargz-snapshotter/blob/v0.4.1/estargz/types.go#L162-L176

Reply:
nevermind my question, I see how it is implemented.

I'll change the metadata to support this use case

Reply:
I've added a patch (completely untested) to support files chunking.

Does the change LGTY?

Reply:
Could we keep the compatibility with stargz/eStargz?

https://github.com/containerd/stargz-snapshotter/blob/v0.4.1/estargz/types.go#L162-L176

Reply:
do you have an example TOC file that uses chunks?

Reply:
I'll drop the last commit and rewrite it to be compatible with stargz/eStargz, an example of TOC file that uses chunks would be helpful to understand better how these fields are used

Reply:
Thanks! 
`sha256:cac892ca7967dfb6bd2433a1c35a5a89c1fe5373725fee80b430c8390f5171cb` of [`ghcr.io/stargz-containers/python:3.9-esgz`](https://github.com/orgs/stargz-containers/packages/container/python/911988) has "chunk" entries (please see also [our doc](https://github.com/containerd/stargz-snapshotter/blob/v0.4.1/docs/pre-converted-images.md) for other pre-converted estargz images, if needed).

And the logic of creating TOC from tar will also help understand it: https://github.com/containerd/stargz-snapshotter/blob/v0.4.1/estargz/estargz.go#L598-L727

---

Excerpt from TOC of `sha256:cac892ca7967dfb6bd2433a1c35a5a89c1fe5373725fee80b430c8390f5171cb`.

```json
		{
			"name": "usr/lib/gcc/x86_64-linux-gnu/8/cc1",
			"type": "reg",
			"size": 25064392,
			"modtime": "2019-04-06T14:44:55Z",
			"mode": 33261,
			"offset": 20405393,
			"NumLink": 0,
			"digest": "sha256:a0993fa7e47d8bc6abf9ec39e10cebae8929ce30b035482c70733e7effa82f39",
			"chunkSize": 4194304,
			"chunkDigest": "sha256:cf1681301b9618990b859a8a74f9fbad1740d7950ace15d6bb03609339aac7a0"
		},
		{
			"name": "usr/lib/gcc/x86_64-linux-gnu/8/cc1",
			"type": "chunk",
			"offset": 22566274,
			"NumLink": 0,
			"chunkOffset": 4194304,
			"chunkSize": 4194304,
			"chunkDigest": "sha256:981f639fb1af5031e9e38ebf2d7dd23bfcc72bbce676f8cd7d694dc12925071b"
		},
		{
			"name": "usr/lib/gcc/x86_64-linux-gnu/8/cc1",
			"type": "chunk",
			"offset": 24537576,
			"NumLink": 0,
			"chunkOffset": 8388608,
			"chunkSize": 4194304,
			"chunkDigest": "sha256:3ed7218ce02b8228487607c0039f076158062c706477d2e3a7cfe626efcf51df"
		},
...
```

Reply:
thanks.

I see that EndOffset is not used.  Is it set implicitly to the next Offset?

Reply:
>  Is it set implicitly to the next Offset?

Yes, it is.

https://github.com/containerd/stargz-snapshotter/blob/v0.4.1/estargz/estargz.go#L417-L421

Reply:
thanks!

Are you fine with the additional metadata I've added?

Reply:
Thanks! LGTM